### PR TITLE
Json improvements

### PR DESCRIPTION
**Pull request by infinisil**

### CODE REVIEW

**Review by lykahb — COMMENTED** [review]
I like that the ByteString is replaced with more precise types 👍

File: src/Crypto/WebAuthn/Metadata/Statement/Types.hs [hunk 1]
```diff
@@ -68,16 +72,30 @@ data MetadataStatement = MetadataStatement
...
   }
-  deriving (Eq, Show, Generic, ToJSON)
+  deriving (Eq, Show, Generic)
+
+-- | An arbitrary and potentially unstable JSON encoding, only intended for
+-- logging purposes. To actually encode and decode structures, use the
+-- "Crypto.WebAuthn.Encoding" modules
+deriving instance ToJSON MetadataStatement
+
+newtype JPEGBytes = JPEGBytes {unJPEGBytes :: BS.ByteString}
```
[the remark this answers is not in the document]
Comment (Infinisil) [reply src/Crypto/WebAuthn/Metadata/Statement/Types.hs:88]: Ah good catch!

**Comment by lykahb** [comment]
@Infinisil Would you like some help to finish this PR? I'd be interested to use webauthn without the orphan instances exported.

File: src/Crypto/WebAuthn/Cose/Internal/Verify.hs [hunk 2]
```diff
@@ -87,21 +90,22 @@ fromX509 key = Left $ "X509 public key algorithm is not supported: " <> Text.pac
 -- 'Cose.PublicKeyWithSignAlg' Returns an error if the signature algorithm
 -- doesn't match. Also returns an error if the signature wasn't valid or for
 -- other errors.
-verify :: Cose.PublicKeyWithSignAlg -> BS.ByteString -> BS.ByteString -> Either Text ()
+-- FIXME: https://w3c.github.io/webauthn/#sctn-signature-attestation-types kind of documents this, but not for all formats. This is notably not really related to COSE, but rather webauthn's own definitions. The spec should be made less ambiguous, file upstream issues and refactor this code
+verify :: Cose.PublicKeyWithSignAlg -> CoseMessage -> CoseSignature -> Either Text ()
```
Comment (infinisil) [inline src/Crypto/WebAuthn/Cose/Internal/Verify.hs:94]: @lykahb I think the need for this PR actually highlights a bigger problem, the one indicated in this FIXME here. The spec notably doesn't really define all binary formats it refers to. Notably I discovered that the message and signature types here are actually a custom format, meaning that they should probably be called `WebAuthnMessage`/`WebAuthnSignature` instead, and this whole module should be moved outside of the Cose module tree.

And the specification is really unclear about this in a bunch of places, I'd like to file an issue in the spec for addressing this.

Comment (infinisil) [reply src/Crypto/WebAuthn/Cose/Internal/Verify.hs:94]: With these holes in the spec, we kind of need to reverse-engineer what the spec really means based on implementations.

Comment (lykahb) [reply src/Crypto/WebAuthn/Cose/Internal/Verify.hs:94]: Yeah, that's a large and not well defined scope. Can we split the changes? The `ByteString` -> `Base16ByteString` that kills the orphan instance, a new type for the PNG data URL, and the comments on the `ToJSON` would be great to have sooner.

**Comment by infinisil** [comment]
I have now created an issue in the webauthn spec highlighting underspecified signature formats: https://github.com/w3c/webauthn/issues/1721

**Comment by lykahb** [comment]
The spec issue is resolved now. I can't tell if the answer makes the signature format more clear, though. 

Is there anything I can help with to get this merged, @infinisil? It seems the only part of this PR that is unclear and related to the spec is the changes for the COSE signatures. So, we could take it out of the PR and instead of replacing `ByteString` with `Cose.CoseSignature`, replace it with `Base16ByteString`. Getting rid of the orphan instances is an easier improvement than COSE and we could do this sooner.

**Comment by infinisil** [comment]
I don't expect to have any time to work on this myself, but I'd be happy to review an updated PR! Splitting this up a bit sounds good too.

**Pull request closed without merging**